File: test/csharp/SpecGateFixtures/Conformance/SumTypes/catch_all.cs
using SpecGate.Annotations;

namespace SpecGateFixtures.Conformance.SumTypes;

/// <summary>
/// Catch-all fallible operation: <c>[SpecException]</c> with no declared types
/// means every thrown exception is the <c>Err</c> arm regardless of type. The
/// two error cases throw DIFFERENT exception types to prove the catch-all is
/// not type-filtered (there is no <c>$fault</c> path for this operation).
/// </summary>
public static class CatchAllOps
{
    /// <summary>Returns <paramref name="x"/> when within [0, 100]; throws otherwise.</summary>
    /// <param name="x">The value to validate (spec input <c>x</c>).</param>
    /// <returns>The value (Ok arm); any thrown exception, of any type, becomes the Err arm.</returns>
    [SpecOperation("require_in_range", Spec = "fixture.catch_all")]
    [SpecException]
    public static int RequireInRange([SpecInput("x")] int x)
    {
        if (x < 0)
        {
            throw new InvalidOperationException("too small");
        }

        if (x > 100)
        {
            throw new FormatException("too big");
        }

        return x;
    }
}

File: test/rust/crates/specgate-fixtures/src/conformance/sum_types/catch_all.rs
// Catch-all fallible op. The C# side uses `[SpecException]` with no declared
// types, so EVERY thrown exception maps to the Err arm regardless of type — the
// two error cases throw DIFFERENT exception types to prove the catch-all is not
// type-filtered (no `$fault` path). The Rust target returns `Err` directly.
use specgate::*;

#[spec_operation("require_in_range", spec = "fixture.catch_all")]
pub fn require_in_range(x: i32) -> Result<i32, String> {
    if x < 0 {
        return Err("too small".to_string());
    }
    if x > 100 {
        return Err("too big".to_string());
    }
    Ok(x)
}
